### DISCUSSION TITLE
Dependency Injection scope logic of Riverpod

### DISCUSSION THREAD

**Discussion by tonnyavery** [rrousselGit/riverpod#2260] [Q&A]
I keep seeing here and there that Riverpod is a good replacement for DI and Service Locator patterns.
How can scope be controlled with Riverpord? I would like to make sure some of my providers are Singletons, while others are atomic functions that just run some processing logic on top of another provider.

I was assuming every provider is Singleton by default. But adding logging entries with `hashCode` I can see that my AsyncNotifierProvider seems to be in multiple instances, so it's `build` method ends up delivering incorrect state when another provider listens.

**Comment by tonnyavery** [comment]
`AutoDisposeAsyncNotifier` constructor gets invoked many times, seemingly every time I use `read()` on it. This results in a new instance and if state is stored inside such async notifier, this results in a new state.

**Reply by rrousselGit** [reply]
Sounds like you have nothing listening to the provider. Therefore it gets disposed

Using "read" doesn't listen to providers.

**Reply by tonnyavery** [reply]
What would be the right way to create a side-effect service or manager?
I have list of notifications provided and need a `manager` that will schedule notifications every time notificationProvider changes.

This is what I did for now.
```
@riverpod
void notificationsManager(NotificationsManagerRef ref) async {
  Logger().d("notificationsManager created $ref");
  ref.listen(notificationConfigsDataStoreProvider, (previous, next) {
    Logger().d("notificationsManager next ${next}");
  });
```

And I use `ref.read(notificationsManagerProvider)` on my Splash Screen `useEffect`, so basically on it's creation.
If I use `watch` there, it would keep my SplashWidget in memory.

How can I keep such `manager` provider alive without hooking it to some Widget instance?

**Reply by tonnyavery** [reply]
Similar question would be for regular, not `manager` providers.
If I have a provider that is essentially a DataStore, so it reads and caches some values from the database or file. I would like to keep such provider alive throughout the lifetime of the app so I don't have to go to the underlying database for data every time I want to read the last state.

**Reply by tonnyavery** [reply]
It looks like with `@riverpod` annotation providers have `.autodispose` by default. Any way to manually control that without switching back to manual provider code?

Also, what would be the implication of using `ref.keepAlive()` if provider that I decided to keep alive was triggered to create by `read` in a Widget, that is already off screen and doesn't display to the user.
